### ISSUE TITLE
CSI Upgrade Tests Failing for hostPath and gcePD

### ISSUE THREAD

**Issue by davidz627** [kubernetes/kubernetes#71228]
Quoted from: https://github.com/kubernetes/kubernetes/issues/65246#issuecomment-439708121

> Failing job
> https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster
> 
> Failing Tests
> 
> [sig-storage] [Serial] CSI Volumes CSI plugin test using CSI driver: gcePD [Serial] should provision storage
> [sig-storage] [Serial] CSI Volumes CSI plugin test using CSI driver: hostPath [Serial] should provision storage
> Error
> https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-cluster/2629
> 
> This is now 1.13 release blocking.

CSI v1.0.0 is only compatible with k8s >1.13
CSI v0.x is only compatible with k8s <1.12

Its probable that we will need to detect what version the k8s cluster is on in the test and deploy the "correct" version of the CSI Drivers for each version

/kind failing-test
/assign
/cc @AishSundar @jberkus @saad-ali @msau42 
/sig storage

**Comment by AishSundar** [comment]
The specific upgrade job these tests are failing on, upgrades the master and nodes to 1.13 and run the old version's tests i.e., in this case runs 1.11 tests against the 1.13 cluster

```If a test job name ends with upgrade-cluster, it means we first upgrade the cluster (i.e. master and nodes) and then run the old test suite with new kubectl.```

More info on upgrade tests at https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md#version-skewed-and-upgrade-testing

**Comment by msau42** [comment]
Do we also have skew tests that only upgrade master, but keeps the nodes on the older version?  If so, we need to detect that case too and use the old 0.3 drivers in that scenario.

**Comment by AishSundar** [comment]
Yes @msau42, thats https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-master

From speaking to @davidz627 we need to update tests in the following jobs

* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-master
* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-master-parallel
* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster
* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster-parallel
* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-master-new-downgrade-cluster
* https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-master-new-downgrade-cluster-parallel

**Comment by liggitt** [comment]
am I understanding correctly that there is no overlap in supported CSI plugin APIs between 1.12 and 1.13?

**Comment by liggitt** [comment]
> the disruptive upgrade path for Kubernetes cluster admins is to remove the old CSI drivers from the cluster before upgrading. Upgrade master and node to k8s v1.13, and wait until master and all nodes are upgraded to 1.13, and then deploy the new CSI 1.0 compatible version of the driver.

Just to explore how disruptive that would be, aren't nodes required to be drained as part of upgrade? if so, when new replicas of those workloads attempt to get scheduled to other nodes, any that use CSI volumes will not be able to run because those volumes could not be attached/mounted, correct?

**Comment by saad-ali** [comment]
> Just to explore how disruptive that would be, aren't nodes required to be drained as part of upgrade? if so, when new replicas of those workloads attempt to get scheduled to other nodes, any that use CSI volumes will not be able to run because those volumes could not be attached/mounted, correct?

Correct, it is suggested that pods be drained from nodes before a node is upgraded. And if a CSI driver is uninstalled before upgrade, then the pods depending on volumes using that driver will remain in pending state (won't attach/mount) until the upgrade is complete and a compatible version of the driver is installed.

**Comment by msau42** [comment]
This might be tricky for pods that are managed by higher level controllers that will automatically restart the pods

**Comment by saad-ali** [comment]
@AishSundar @liggitt I spoke with @thockin. He is not happy with the backwards comparability break. He wants us to look in to what it will take to support both 0.x and 1.0. I will look in to that. And update this thread on feasibility. Until this is resolved consider this a release blocking issue.

/priority critical-urgent 
/milestone v1.13

**Comment by AishSundar** [comment]
@saad-ali thanks for the update, will watch this thread.

**Comment by AishSundar** [comment]
/reopen

Reopening until we get green CI. @jberkus ^^

**Comment by jberkus** [comment]
We're getting a lot of flakes across the longer-running tests today, so at this point we don't know if things are passing or not.

**Comment by liggitt** [comment]
the CSI tests in question turned green after #71314 was merged:

https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster&include-filter-by-regex=.*CSI.*&width=20

This particular issue can be closed

**Comment by jberkus** [comment]
@liggitt @saad-ali We're still getting various CSI failures in new-master-upgrade-parallel:

https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster-new-parallel&include-filter-by-regex=.*CSI.*&width=20

Are those unrelated to this issue?  They seem to be all subpath failures.

**Comment by liggitt** [comment]
> Are those unrelated to this issue?

Flakes are unrelated to this issue. This issue was fixing a hard fail.

**Comment by jberkus** [comment]
OK, closing this and opening a new issue for the flakes.

/close
